Rewrite core of PyMC3 (upcoming v4) to rely on aesara (among many other improvements)

twiecki opened the pull request.

brandonwillard committed:
Initial refactoring for RandomVariable use in Model, step methods, and basic dists

These changes can be summarized as follows:
- `Model` objects now track fully functional Theano graphs that represent all
relationships between random and "deterministic" variables.  These graphs are
called these "sample-space" graphs.  `Model.unobserved_RVs`, `Model.basic_RVs`,
`Model.free_RVs`, and `Model.observed_RVs` contain these
graphs (i.e. `TensorVariable`s), which are generated by `RandomVariable` `Op`s.
- For each random variable, there is now a corresponding "measure-space"
variable (i.e. a `TensorVariable` that corresponds to said variable in a
log-likelihood graph).  These variables are available as `rv_var.tag.value_var`,
for each random variable `rv_var`, or via `Model.vars`.
- Log-likelihood (i.e. measure-space) graphs are now created for individual
random variables by way of the generic functions `logpt`, `logcdf`,
`logp_nojac`, and `logpt_sum` in `pymc3.distributions`.
- Numerous uses of concrete shape information stemming from `Model`
objects (e.g. `Model.size`) have been removed/refactored.
- Use of `FreeRV`, `ObservedRV`, `MultiObservedRV`, and `TransformedRV` has been
deprecated.  The information previously stored in these classes is now tracked
using `TensorVariable.tag`, and log-likelihoods are generated using the
aforementioned `log*` generic functions.

brandonwillard committed:
Temporarily disable CI tests

brandonwillard committed:
Remove shape dependencies from DictToArrayBijection

This commit changes `DictToArrayBijection` so that it returns a `RaveledVars`
datatype that contains the original raveled and concatenated vector along with
the information needed to revert it back to dictionay/variables form.

Simply put, the variables-to-single-vector mapping steps have been pushed away
from the model object and its symbolic terms and closer to the (sampling)
processes that produce and work with `ndarray` values for said terms.  In doing
so, we can operate under fewer unnecessarily strong assumptions (e.g. that the
shapes of each term are static and equal to the initial test points), and let
the sampling processes that require vector-only steps deal with any changes in
the mappings.

brandonwillard committed:
Rename Model.ndim to Model.size

This value was not representative of its name.

brandonwillard committed:
Update competence methods to work with RandomVariables

brandonwillard committed:
Implement naive RandomVariable-based posterior predictive sampling

The approach currently being used is rather inefficient.  Instead, we should
change the `size` parameters for `RandomVariable` terms in the sample-space
graph(s) so that they match arrays of the inputs in the trace and the desired
number of output samples.  This would allow the compiled graph to vectorize
operations (when it can) and sample variables more efficiently in large batches.

brandonwillard committed:
Remove newly deprecated classes and functions

Classes and functions removed:
- PyMC3Variable
- ObservedRV
- FreeRV
- MultiObservedRV
- TransformedRV
- ArrayOrdering
- VarMap
- DataMap
- _DrawValuesContext
- _DrawValuesContextBlocker
- is_fast_drawable
- _compile_theano_function
- vectorize_theano_function
- get_vectorize_signature
- _draw_value
- draw_values
- generate_samples
- fast_sample_posterior_predictive

Modules removed:
- pymc3.distributions.posterior_predictive
- pymc3.tests.test_random

fonnesbeck committed:
Removed redundant bound in Wald distribution

brandonwillard committed:
Refactor tests for compatibility with logp dispatch and RandomVariables

OriolAbril committed:
Port InferenceData conversion code

michaelosthege committed:
Apply easy fixes to get tests to pass or xfail

michaelosthege committed:
Allow ignoring tests files, but print a warning about it

michaelosthege committed:
Make XPASSing tests fail, so we'll know when something is fixed

Closes #4516

michaelosthege committed:
Ignore tests that are completely broken, but run the others

brandonwillard committed:
Reinstate log-likelihood transforms

michaelosthege commented on pymc3/blocking.py:
```diff
@@ -18,245 +18,80 @@
 
 
-# TODO Classes and methods need to be fully documented.
+class DictToArrayBijection:
```
Well condensed!

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -11,6 +11,373 @@
+]
+
+
+def _get_scaling(total_size, shape, ndim):
```
Adding the type hints wouldn't hurt - they are already in the docstring

michaelosthege commented on pymc3/distributions/__init__.py:
```diff
@@ -11,6 +11,373 @@
+        raise TypeError(
+            "Unrecognized `total_size` type, expected int or list of ints, got %r" % total_size
+        )
+    return aet.as_tensor(floatX(coef))
```
If I'm not mistaken the `floatX` here is not necessary

ricardoV94 committed:
Strip observed in prior predictive

ricardoV94 committed:
Fix values in `test_ignores_observed`

brandonwillard committed:
Use name from Variable.name in pymc3.util.get_var_name

brandonwillard committed:
Add fallback for missing value_var in BaseTrace

brandonwillard committed:
Remove remaining v3 sampling code

brandonwillard committed:
Change logp_transform argument to the entire random variable

brandonwillard committed:
Remove logpt transformed option

brandonwillard committed:
Implement transform for Dirichlet

brandonwillard committed:
Always use the value var to initially build the log-likelihood

brandonwillard committed:
Register GammaRV to PositiveContinuous class

brandonwillard committed:
Add an option for negative support assertions in Normal and Gamma classes

brandonwillard committed:
Make logp testing work with transformed values

brandonwillard committed:
Disable asserts during logp invalid range tests

brandonwillard committed:
Update xfails in pymc3.tests.test_distributions

brandonwillard committed:
Fix Categorical logp implementation

brandonwillard committed:
Use the value var's values directly in logpt

brandonwillard committed:
Do not assume values are Applys in pymc3.tests.test_distributions

kc611 committed:
Refactored distributions in pymc.distributions.continuous

brandonwillard committed:
Simplify the new Distribution interface and convert a few more

brandonwillard committed:
Use xfail mark in pymc3.tests.test_distributions_timeseries

ricardoV94 commented:
In case this helps:

1. Since https://github.com/pymc-devs/pymc3/commit/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d, prior sampling returns the same value for every draw.

```python
with pm.Model() as m:
    x = pm.Normal('x', 0, 1)
    prior = pm.sample_prior_predictive(10)
    print(prior)
    
{'x': array([0.019066, 0.019066, 0.019066, 0.019066, 0.019066, 0.019066,
       0.019066, 0.019066, 0.019066, 0.019066])}
```

2. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?

brandonwillard commented:
> 1. Since [a80cf9a](https://github.com/pymc-devs/pymc3/commit/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d), prior sampling returns the same value for every draw.

Yes, that's because it's using the same RNG seed every time.  The reason why: `RandomVariable.inplace == False`.  [This code](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/distribution.py#L87) was supposed to take care of that.

brandonwillard commented:
> 1. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?

Is there an example of this, or a test that's failing?

ricardoV94 commented:
> > 1. For the HalfCauchy (and maybe other distributions) there may be a disconnect between the logp and the random arguments. Our logp considers only the scale argument, but the random method (from scipy) takes loc as well and that's the first optional argument. I don't think there is yet any logic to connect the arguments between the the random and logp, right?
> 
> Is there an example of this, or a test that's failing?

The random tests are still disabled, so this wouldn't show up right? I wanted to check with the prior predictive sampling but right now that's also unusable. 

There was an issue in the logp of the HalfCauchy that lead to the addition of an unused parameter (not sure if related):
https://github.com/pymc-devs/pymc3/pull/4508/files/3db730e1b632e0fa023cd181c60d19de4aab6e48#r594550615

ricardoV94 commented:
Managed to do it with `.dist`

```python
import pymc3 as pm
import scipy.stats as st

pymc_samples = pm.HalfCauchy.dist(beta=.01, size=10_000).eval()
print(pymc_samples.mean(), pymc_samples.std())
# (6.647935041418325, 81.64505597965584)

# Shoud match this 
scipy_samples = st.halfcauchy(loc=0, scale=.01).rvs(10_000)
print(scipy_samples.mean(), scipy_samples.std())
# (0.049328030173446065, 0.3650001369075893)

# But matches this
scipy_samples = st.halfcauchy(loc=0.01, scale=1).rvs(10_000)
print(scipy_samples.mean(), scipy_samples.std())
# (5.537612320455185, 73.96359661595363)

# V3
pymc_samples = pm.HalfCauchy.dist(beta=.01, shape=10_000).random()      
print(pymc_samples.mean(), pymc_samples.std())                          
# (0.05829908302484347, 1.122931711761449)
```

The HalfCauchy is notoriously hard to evaluate, but with these parameters it seems reliable.

brandonwillard committed:
Create class after setting the class dictionary

brandonwillard commented:
> The random tests are still disabled, so this wouldn't show up right?

The tests in `pymc3.tests.test_distributions` perform some sampling and those are running for all the converted `Distribution`s; otherwise, I believe you're talking about `pymc3.tests.test_distributions_random`.  Those tests are *not* enabled because they're redundant (for the `RandomVariable`s that already exist in Aesara, at least).

>  I wanted to check with the prior predictive sampling but right now that's also unusable.

It's not unusable; you only need to set/change the seed between calls to `pm.sample_*_predictive`.  Regardless, I've pushed a commit that correctly sets `RandomVariable.inplace = True`, so you shouldn't need to do that anymore.

ricardoV94 commented:
Thanks Brandon. 

I didn't mean it's unusable. I just didn't know how to make use of it :b


I confirmed my suspicion about the Half Cauchy though. There is a mismatch between the old logp and the new randomOp, as the later expects two arguments, while the former expects only one. 

That's why the logp test would fail if the unused `alpha` argument was not added here: https://github.com/pymc-devs/pymc3/blob/a80cf9ac7dec48cc71a2924362884a4d8cc6e30d/pymc3/distributions/continuous.py#L2276

It's not too difficult to add this loc parameter to the logp, but just wanted to make sure we want to do that. Do we want to adapt to the number of arguments expected in aesara or just customize our own random ops in terms of the arguments we want to use?

brandonwillard commented:
> It's not too difficult to add this loc parameter to the logp, but just wanted to make sure we want to do that. Do we want to adapt to the number of arguments expected in aesara or just customize our own random ops in terms of the arguments we want to use?

If an adaptation is something as simple as inverting a parameter, then we'll always want to do that, but this case sounds like we'll need to do the latter (i.e. customize a `RandomVariable`).  There's now an [example of such a customization for `Multinomial`](https://github.com/pymc-devs/pymc3/blob/v4/pymc3/distributions/multivariate.py#L434).

ricardoV94 committed:
Fix HalfCauchy/HalfCauchyRV parameterization

ricardoV94 commented:
On another front:

```python
import pymc3 as pm
with pm.Model() as m:
    x = pm.Uniform('x', 0, 1)
    y = pm.Uniform('y', 0, x)
    
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.83990034)
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.09721045)
m.logp({'x_interval': 0, 'y_interval': 0})
# array(-2.23865389)
```

Also it ignores when I set transform=None, and creates the interval variables anyway.

michaelosthege commented:
@ricardoV94 were the trailing `__` on the transformed variables removed?

On `master`:
```python
>>> m.logp({'x_interval': 0, 'y_interval': 0})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\osthege\Repos\pymc3-dev\pymc3\model.py", line 1565, in __call__
    return self.f(**point)
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3-dev\lib\site-packages\aesara\compile\function\types.py", line 956, in __call__
    raise TypeError(
TypeError: Missing required input: x_interval__ ~ TransformedDistribution
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
>>> m.logp({'x_interval__': 0, 'y_interval__': 0})
array(-2.77258872)
```

ricardoV94 commented:
Yes they were removed for the time being.

brandonwillard commented:
> Also it ignores when I set transform=None, and creates the interval variables anyway.

I'll add fixes for those in a minute.

brandonwillard committed:
Make transform objects stateless

ricardoV94 committed:
Fix shape logic in test_distributions_random and skip tests covered by Aesara

ricardoV94 committed:
Re-enable test_distributions_random

ricardoV94 committed:
Refactor Beta to use custom rng_fn clipped_beta_rv

ricardoV94 committed:
Fix HalfNormal/HalfNormalRV parameterization

brandonwillard committed:
Add tests for two important open logpt and Model issues

brandonwillard committed:
Add non_sequences to uses of Scan Op

This make `aesara.graph.basic.clone_replace` work correctly when `Scan`s are
included in a graph.

brandonwillard committed:
Replace Observed Op with tag.observations

brandonwillard committed:
Add missing imports to pymc3.step_methods.gibbs

brandonwillard committed:
Comment out unused moments

brandonwillard committed:
Make logpt work correctly for nested models and transforms

brandonwillard committed:
Disable use of Arviz in pymc3.tests.test_data_container

brandonwillard committed:
Set model seed correctly in pymc3.tests.test_ndarray_backend

brandonwillard committed:
Prevent Model from turning on test value computations

brandonwillard committed:
Make Metropolis, Slice, PGBART, MetropolisMLDA use point values

brandonwillard committed:
Re-enable disabled tests

brandonwillard committed:
Use value vars to determine steps

brandonwillard committed:
Use size instead of shape in pymc3.tests.sampler_fixtures

brandonwillard committed:
Set model-level RandomVariable seeds during sampling

brandonwillard committed:
Check shapes by evaluating graph with start values

brandonwillard committed:
Fix logpt so that transforms are always applied, when enabled

brandonwillard committed:
Set default transform for Dirichlet

brandonwillard committed:
Normalize Multinomial argument

brandonwillard committed:
Use no_transform_object in Distribution.__new__

brandonwillard committed:
Fix Interval.jacobian_det

brandonwillard committed:
Fix Stickbreaking scalar condition

brandonwillard committed:
Make Model.test_point generation transform existing RV test value

brandonwillard committed:
Add transformed value variables to Model.named_vars

brandonwillard committed:
Remove DeterministicWrapper from Deterministic

brandonwillard committed:
Make sure sample_posterior_predictive doesn't use trace values for sampled vars

brandonwillard committed:
Set seed after loading trace in TestSaveLoad

brandonwillard committed:
Make v4 compatibility changes to pymc3.tests.test_sampling

brandonwillard committed:
Make pymc3.tests.test_transforms work with None RV variables

brandonwillard committed:
Make find_MAP work with RaveledVars

brandonwillard committed:
Make sure start values are NumPy arrays

brandonwillard committed:
Raise NotImplementedError in Group.__init__

brandonwillard committed:
Add type hints to astep methods

brandonwillard committed:
Use untransformed samples and xfail Arviz tests in BaseSampler

brandonwillard committed:
Make sure forward transformed input is a TensorVariable in TestMatchesScipy

brandonwillard committed:
Adjust flaky last-digit numerical requirements in TestMatchesScipy

brandonwillard committed:
Fix MvNormal quaddist_matrix parameter order

brandonwillard committed:
Enable MvNormal tests in test_distributions

brandonwillard committed:
Factor out parameter pre-processing in TestMatchesScipy

brandonwillard committed:
Apply recent xfail updates from master branch

brandonwillard committed:
Prevent dtype conversion in Aesara during testing to avoid a bug

This change avoids the bug fixed in
https://github.com/pymc-devs/aesara/pull/348.

brandonwillard committed:
Fix NegativeBinomial parameterization and enable its tests

brandonwillard committed:
Prevent SciPy error by using float64 point in test_dirichlet_with_batch_shapes

brandonwillard committed:
Disable dims, default_dims, and index_origin options until arviz > v0.11.2

brandonwillard committed:
Create extract_obs_data function

brandonwillard committed:
Re-enable Arviz tests in pymc3.tests.test_sampling

brandonwillard committed:
Do not use shared variables as inputs during prior/posterior sampling

OriolAbril committed:
add workaround for data groups until next arviz release

OriolAbril committed:
activate arviz compat tests

brandonwillard committed:
Remove redundant coords test

brandonwillard committed:
Re-enable Arviz parts of sampler fixtures

brandonwillard committed:
Re-enable v4 xfails in pymc3.distributions.dist_math

brandonwillard committed:
Remove deprecated default values tests

brandonwillard committed:
Re-enable pymc3.tests.test_modelcontext

brandonwillard committed:
Fix extra_vars in call to ValueGradFunction from Model

brandonwillard committed:
Re-enable some tests in pymc3.tests.test_model

brandonwillard committed:
Remove Model.size

brandonwillard committed:
Always set model.default_rng in pymc3.sampling.sample

brandonwillard committed:
Change shape to size in pymc3.tests.test_step

brandonwillard committed:
Mark brittle TestStepMethods.test_sample_exact test as xfail

brandonwillard committed:
Remove incorrect size for multivariate distributions in pymc3.tests.models

brandonwillard committed:
Change shape to size in pymc3.tests.test_aesaraf

brandonwillard committed:
Make minor v4 updates to test_car_logp

brandonwillard committed:
Re-enable more sampler-based tests and selectively xfail others

brandonwillard committed:
Update xfail reason in pymc3.tests.test_examples

michaelosthege reviewed (commented):
Made these comments a while back, but forgot to hit submit..

brandonwillard committed:
Set a seed for a flaky pymc3.tests.test_sampling test

brandonwillard committed:
Rename Model.vars to Model.value_vars

brandonwillard committed:
Reintroduce the old Model.vars with a deprecation warning

brandonwillard committed:
Add value/rv variable maps to Model

brandonwillard committed:
Change no_transform_object sentinel to UNSET in pymc3.util

brandonwillard committed:
Move logp code into pymc3.distributions.logp and graph functions into pymc3.aesaraf

brandonwillard committed:
Remove build_named_node_tree and its helper functions

brandonwillard committed:
Move treelist and treedict to pymc3.util

brandonwillard committed:
Remove pymc3.model._walk_up_rv

brandonwillard committed:
Remove pymc3.model.as_iterargs

brandonwillard committed:
Move pymc3.model.all_continuous to pymc3.sampling

brandonwillard committed:
Move pymc3.model.pandas_to_array to pymc3.aesaraf

brandonwillard committed:
Rename pymc3.tests.test_model_utils to test_glm_utils

brandonwillard committed:
Rename pymc3.tests.test_models_linear to test_glm_linear

brandonwillard committed:
Add xfail to pymc3.tests.test_variational_inference

brandonwillard committed:
Remove remaining uses of Distribution.default

brandonwillard committed:
Remove Distribution.get_test_val and Distribution.getattr_value

brandonwillard committed:
Rename Model.test_point to Model.initial_point

brandonwillard committed:
Rename Model.check_test_point to Model.point_logps

brandonwillard committed:
Make pymc3.util.update_start_vals and check_start_vals Model methods

brandonwillard committed:
Re-enable pymc3.tests.test_util

brandonwillard committed:
Fix issue with single variable input to BlockedStep

brandonwillard committed:
Make GradientSharedStep a subclass of ArrayStepShared

brandonwillard committed:
Make initial_replacements argument in rvs_to_value_vars optional

brandonwillard committed:
Rename rv_ancestors to walk_model and add tests

brandonwillard committed:
Merge and improve apply_transforms and rvs_to_value_vars

brandonwillard committed:
Fix a NumPy non-tuple indexing deprecation warning in test_aesaraf

brandonwillard committed:
Convert and transform potentials in the log-likelihood

brandonwillard committed:
Finish refactoring BlockedStep.competence implementations

brandonwillard committed:
Fix RaveledVars and size-related issues in Metropolis and MLDA samplers

brandonwillard committed:
Seed flaky TestSamplePPC.test_deterministic_of_observed_modified_interface

brandonwillard committed:
Re-enable step method tests in pymc3.tests.test_step

brandonwillard committed:
Add a test module for pymc3.distributions.logp

brandonwillard committed:
Implement log-likelihoods for *IncSubtensor* Ops

fonnesbeck committed:
Converted Gumbel distribution to v4

fonnesbeck committed:
Ran pre-commit scripts

brandonwillard committed:
Use the Subtensor-ed result as the observed random variable

brandonwillard committed:
Remove unnecessary module pymc3.tests.test_model_helpers

brandonwillard committed:
Update Arviz conversion for new missing values approach

brandonwillard committed:
Update tests for new missing values approach

brandonwillard committed:
Allow empty RaveledVars results in DictToArrayBijection.map

fonnesbeck committed:
Converted Pareto distribution to v4

brandonwillard committed:
Allow dict of value vars in logp signatures and implement Subtensor logp

ricardoV94 commented:
@brandonwillard I poked around after you made the missing data changes to check whether we are auto-transforming the imputed variables or not, but I am finding some issues before even getting there. 

In V3:
```python
with pm.Model() as m:
  x = pm.Gamma('x', 1, 1, observed=np.array([1, 1, np.nan, 1]))
print(m.vars)  # [x_missing ~ NoDistribution]
print(m.logp({'x_missing': [1]}))  # 4.0
```
in V4:
```python
with pm.Model() as m:
  x = pm.Gamma('x', 1, 1, observed=np.array([1, 1, np.nan, 1]))
# Not sure if we should add `x` to m.vars...
print(m.vars)  # [ ]

# Evaluation is stochastic
print(m.logp({'x_missing': [1]}))  # -4.487674097730719
print(m.logp({'x_missing': [1]}))  # -3.195617080537621

# I can put any variable name, shouldn't this raise an Error regardless?
print(m.logp({'z': [1]}))  # -3.165731061987242
```

The random Gamma variable is still in the logp graph:
```python
print(aesara.dprint(m.logpt))
Sum{acc_dtype=float64} [id A] '__logp'   
 |MakeVector{dtype='float64'} [id B] ''   
   |Sum{acc_dtype=float64} [id C] ''   
     |Elemwise{mul,no_inplace} [id D] '__logp_x'   
       |Sum{acc_dtype=float64} [id E] ''   
       | |Elemwise{switch,no_inplace} [id F] ''   
       |   |Elemwise{mul,no_inplace} [id G] ''   
       |   | |Elemwise{mul,no_inplace} [id H] ''   
       |   | | |Elemwise{mul,no_inplace} [id I] ''   
       |   | | | |InplaceDimShuffle{x} [id J] ''   
       |   | | | | |TensorConstant{1} [id K]
       |   | | | |Elemwise{mul,no_inplace} [id L] ''   
       |   | | |   |InplaceDimShuffle{x} [id M] ''   
       |   | | |   | |TensorConstant{1} [id N]
       |   | | |   |Elemwise{ge,no_inplace} [id O] ''   
       |   | | |     |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   | | |     | |DisconnectedGrad [id Q] ''   
       |   | | |     | | |halfnormal_rv.1 [id R] 'x'   
       |   | | |     | |   |RandomStateSharedVariable(<RandomState(MT19937) at 0x7F4779136140>) [id S]
       |   | | |     | |   |TensorConstant{(1,) of 4} [id T]
       |   | | |     | |   |TensorConstant{11} [id U]
       |   | | |     | |   |TensorConstant{1.0} [id V]
       |   | | |     | |   |Elemwise{true_div,no_inplace} [id W] ''   
       |   | | |     | |     |TensorConstant{1.0} [id X]
       |   | | |     | |     |Elemwise{inv,no_inplace} [id Y] ''   
       |   | | |     | |       |TensorConstant{1.0} [id Z]
       |   | | |     | |TensorConstant{(3,) of 1.0} [id BA]
       |   | | |     | |TensorConstant{[ True  Tr..lse  True]} [id BB]
       |   | | |     |InplaceDimShuffle{x} [id BC] ''   
       |   | | |       |TensorConstant{0} [id BD]
       |   | | |InplaceDimShuffle{x} [id BE] ''   
       |   | |   |Elemwise{mul,no_inplace} [id BF] ''   
       |   | |     |TensorConstant{1} [id BG]
       |   | |     |Elemwise{gt,no_inplace} [id BH] ''   
       |   | |       |TensorConstant{1.0} [id V]
       |   | |       |TensorConstant{0} [id BI]
       |   | |InplaceDimShuffle{x} [id BJ] ''   
       |   |   |Elemwise{mul,no_inplace} [id BK] ''   
       |   |     |TensorConstant{1} [id BL]
       |   |     |Elemwise{gt,no_inplace} [id BM] ''   
       |   |       |Elemwise{true_div,no_inplace} [id BN] ''   
       |   |       | |TensorConstant{1.0} [id X]
       |   |       | |Elemwise{inv,no_inplace} [id BO] ''   
       |   |       |   |TensorConstant{1.0} [id Z]
       |   |       |TensorConstant{0} [id BP]
       |   |Elemwise{add,no_inplace} [id BQ] ''   
       |   | |Elemwise{sub,no_inplace} [id BR] ''   
       |   | | |InplaceDimShuffle{x} [id BS] ''   
       |   | | | |Elemwise{add,no_inplace} [id BT] ''   
       |   | | |   |Elemwise{neg,no_inplace} [id BU] ''   
       |   | | |   | |gammaln [id BV] ''   
       |   | | |   |   |TensorConstant{1.0} [id V]
       |   | | |   |Elemwise{switch,no_inplace} [id BW] ''   
       |   | | |     |Elemwise{eq,no_inplace} [id BX] ''   
       |   | | |     | |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | | |     | |TensorConstant{0} [id BY]
       |   | | |     |Elemwise{switch,no_inplace} [id BZ] ''   
       |   | | |     | |Elemwise{eq,no_inplace} [id CA] ''   
       |   | | |     | | |TensorConstant{1.0} [id V]
       |   | | |     | | |TensorConstant{0} [id CB]
       |   | | |     | |TensorConstant{0.0} [id CC]
       |   | | |     | |TensorConstant{-inf} [id CD]
       |   | | |     |Elemwise{mul,no_inplace} [id CE] ''   
       |   | | |       |TensorConstant{1.0} [id V]
       |   | | |       |Elemwise{log,no_inplace} [id CF] ''   
       |   | | |         |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | | |Elemwise{mul,no_inplace} [id CG] ''   
       |   | |   |InplaceDimShuffle{x} [id CH] ''   
       |   | |   | |Elemwise{true_div,no_inplace} [id BN] ''   
       |   | |   |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   | |Elemwise{switch,no_inplace} [id CI] ''   
       |   |   |Elemwise{eq,no_inplace} [id CJ] ''   
       |   |   | |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   |   | |InplaceDimShuffle{x} [id CK] ''   
       |   |   |   |TensorConstant{0} [id CL]
       |   |   |InplaceDimShuffle{x} [id CM] ''   
       |   |   | |Elemwise{switch,no_inplace} [id CN] ''   
       |   |   |   |Elemwise{eq,no_inplace} [id CO] ''   
       |   |   |   | |Elemwise{sub,no_inplace} [id CP] ''   
       |   |   |   | | |TensorConstant{1.0} [id V]
       |   |   |   | | |TensorConstant{1} [id CQ]
       |   |   |   | |TensorConstant{0} [id CR]
       |   |   |   |TensorConstant{0.0} [id CS]
       |   |   |   |TensorConstant{-inf} [id CT]
       |   |   |Elemwise{mul,no_inplace} [id CU] ''   
       |   |     |InplaceDimShuffle{x} [id CV] ''   
       |   |     | |Elemwise{sub,no_inplace} [id CP] ''   
       |   |     |Elemwise{log,no_inplace} [id CW] ''   
       |   |       |AdvancedIncSubtensor{inplace=False,  set_instead_of_inc=True} [id P] ''   
       |   |InplaceDimShuffle{x} [id CX] ''   
       |     |TensorConstant{-inf} [id CY]
       |TensorConstant{1.0} [id CZ]
```

brandonwillard commented:
> @brandonwillard I poked around after you made the missing data changes to check whether we are auto-transforming the imputed variables or not, but I am finding some issues before even getting there.

Yes, exactly, there are no model-level random variables added to the graph for missing values.  The missing data look just like normal data to the model, and they don't need to be transformed.  Unless we're going to compute gradients in those missing terms and attempt to estimate them, why would we want to transform them.

Also, the evaluation should necessarily be stochastic if we're imputing missing values via sampling.  We might need to make sure that the RNG states are cloned between multiple sampler steps, though, or that the same `*Subtensor*`-updated observation array is used by all of them.

ricardoV94 commented:
I think we indeed used to infer these during sampling, using NUTS as usual. Wasn't that the point of having them?

twiecki commented:
> I think we indeed used to infer these during sampling, using NUTS as usual. Wasn't that the point of having them?

Yes, that's right.

ricardoV94 commented:
Okay I was not aware that was a feature and not a bug. Are we sure we want, by default, to impute values by random sampling instead of NUTS as before?

This also means we can only impute values for distributions with random methods.

brandonwillard commented:
Yes, for variables without a sampling method we're going to need to add them to the model to be sampled themselves.

I believe we can do that just as easily by changing the `RandomVariable` part of the `*IncSubtensor*` to be the array of unobserved terms instead, and simply add those terms to the model variables.

ricardoV94 commented:
Alright so the question then is what to use as the default, even if both ways are possible right?

brandonwillard commented:
> Alright so the question then is what to use as the default, even if both ways are possible right?

Sampling the current way&mdash;i.e. some form of "exact" sampling&mdash;will be a lot cheaper, more "accurate", and faster in some cases, but definitely not all.

brandonwillard commented:
N.B.: You know, this missing data situation is one of those cases in which having one Aesara graph that performs all the sampling steps would be a real advantage, because it would remove all the out-of-graph work required to sync and/or recompute values (especially exactly sampled missing data) across different sampler steps.  

For that matter, there could easily be entire (sub)graphs that we're currently recomputing between steps.  Also, performing in-place updates between the same steps is so much more difficult with the current approach, because we have to implement it by hand, whereas such things could be automatically applied as a single graph&mdash;even with the existing optimizations!

ricardoV94 commented:
Less important than this, from my comment above shouldn't the `model.logp` have raised an error when passing a dictionary with a variable not in the graph?

brandonwillard commented:
> Less important than this, from my comment above shouldn't the `model.logp` have raised an error when passing a dictionary with a variable not in the graph?

No, no reason to do that.  Plus, the implementation would become unnecessarily more difficult that way.

ricardoV94 committed:
Fix exponential and gamma logp / random link (#4576)

brandonwillard committed:
Add missing values as unobserved random variables and estimate them during MCMC

brandonwillard committed:
Add a starting point option to DictToArrayBijection.rmap

brandonwillard committed:
Make sure shared variables have correct broadcasting in ValueGradFunction

brandonwillard committed:
Fix RaveledVars usage in BinaryGibbsMetropolis.astep

brandonwillard committed:
Use ABC abstract methods in step method interface

brandonwillard committed:
Format find_MAP docstring

brandonwillard committed:
Enable Arviz tests

michaelosthege committed:
Compute the size of scalars for size expansions in pymc3.aesaraf.change_rv_size

larryshamalama committed:
Fixed grammar typos: "a aesara" to "an Aesara" (take 3) (#4624)

* Fixed Aesara related spelling in documentation

ckrapu committed:
Fix sequential backend (#4584)

* Fixing error with sequential chain sampling for NDArray backend (closes #4568)

michaelosthege committed:
Remove nonexistent attribute from __all__ list (#4627)

ricardoV94 committed:
Fix Uniform logp regression from #4541

brandonwillard committed:
Use optimized graph for JAX conversion

ricardoV94 committed:
Refactor DiscreteWeibull (#4615)

brandonwillard committed:
Create a NumPyro sampler Op for better JAX backend integration

ricardoV94 committed:
Add unobserved_value_vars property to Model

Fixes pm.Deterministics still using rvs during pm.Sample()
Fixes find_MAP() failing when pm.Deterministics were present in model

ricardoV94 committed:
Add auto_deterministics list to Model

Ensures that when missing variables are present in the model, the automatic deterministic (x_observed + x_missing) only appears in predictive sampling and not normal sampling.
Fixes `x` missing from prior_predictive when missing values were present (only `x_missing` was present)

michaelosthege committed:
Allow unnamed (None) dims and undefined (None) coord values

Also refactor into properties to add docstrings and type annotations.
And no longer allow InferenceData conversion without a Model on stack.

Co-authored-by: Oriol Abril Pla <oriol.abril.pla@gmail.com>

michaelosthege committed:
Allow parametrization through either shape, dims or size

Always treats `size` as being in addition to dimensions implied by RV parameters.

All resizing beyond parameter-implied dimensionality is done from:
- `shape` or `size` in `Distribution.dist()`
- `dims` or `observed` in `Distribution.__new__`
and only in those two places.

Closes #4552.

michaelosthege committed:
Automatically add SpecifyShape Op when full-length shape is given

ricardoV94 committed:
Fix missing fstrings

Spaak committed:
remove obsolete multiplication with 1.0 in get_sigma_tau

MarcoGorelli committed:
fixup stacklevel

matteo-pallini committed:
V4 update test framework for distributions random method (#4608)

* Update tests following distributions refactoring

The distributions refactoring moves the random variable sampling to
aesara. This relies on numpy and scipy random variables implementation.
So, now the only thing we care about testing is that the parametrization
on the PyMC side is sendible given the one on the Aesara side
(effectively the numpy/scipy one)

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Change tests for more refactored distributions.

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Change tests for refactored distributions

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Remove tests for random variable samples shape and size

Most of the random variable logic has been moved to aesara, as well as
most of the relative tests. More details can be found on issue #4554

* Fix test for half cauchy, renmae mv normal tests and add test for
Bernoulli

* Add test checking PyMC samples match the aesara ones

Also mark test_categorical as expected to fail due to bug on aesara
side. The bug is going to be fixed with 2.0.5 release, so we need to
bump the version for categorical and the test to pass.

* Move Aesara to 2.0.5 to include Gumbel distribution

* Enamble exponential and gamma tests following bug-fix

* Enable categorical test following aesara version bump to 2.0.5 and relative bug-fix

* Few small cosmetic changes:
- replace list of tuples with dict
- rename 1 method
- move pymc_dist as first argument in function call
- replace list(params) with params.copy()

* Remove redundant tests

* Further refactoring

The refactoring should make it possible testing both the distribution
parametrization and sampled values according to need, as well as any
other future test. More details on PR #4608

* Add size tests to new rv testing framework

* Add tests for multivariate and for univariate multi-parameters

* remove test already covered in aesara

* fix few names

* Remove "distribution" from test class names

* Add discrete Weibull, improve Beta and some minor refactoring

* Fix typos in checks naming and add sanity check

Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

larryshamalama committed:
Fixing np.bool Depreciation warnings: 3 files (#4673)

* :turtle: Fixed documentation

* :turtle: Minor fix on top of grammar fixed in docs

* Weibull RandomVariable refactoring: an example

* Revert "Weibull RandomVariable refactoring: an example"

This reverts commit 685a06554195eeb73b07fceea90fb783187990cb.

* Fixing np.bool depreciation warnings: 3 files

* Changed np.bool_ to bool (+ fixed other places)

ricardoV94 committed:
Refactor DiscreteUniform

ricardoV94 committed:
Refactor Constant

ricardoV94 committed:
Refactor OrderedLogistic

ricardoV94 committed:
Refactor OrderedProbit

ricardoV94 committed:
Add missing discrete distributions to API rst

ricardoV94 committed:
Refactor ZeroInflatedPoisson

ricardoV94 committed:
Refactor ZeroInflatedBinomial

ricardoV94 committed:
Refactor ZeroInflatedNegativeBinomial

ricardoV94 committed:
Update several test xfails

ricardoV94 commented:
Sampling is failing when passing a dictionary for starting point with values for "untransformed" variables. The old test `TestDisasterModel` in `test_examples.py` that was marked as xfail shows this well:

```python
@pytest.mark.xfail(
    reason="_check_start_shape fails with start dictionary"
    # condition=(aesara.config.floatX == "float32"), reason="Fails on float32"
)
class TestDisasterModel(SeededTest):
    @pytest.mark.xfail(condition=(aesara.config.floatX == "float32"), reason="Fails on float32")
    # Time series of recorded coal mining disasters in the UK from 1851 to 1962
    def test_disaster_model(self):
        model = build_disaster_model(masked=False)
        with model:
            # Initial values for stochastic nodes
            start = {"early_mean": 2, "late_mean": 3.0}
            # Use slice sampler for means (other variables auto-selected)
            step = pm.Slice([model["early_mean_log__"], model["late_mean_log__"]])
            tr = pm.sample(500, tune=50, start=start, step=step, chains=2)
```
The problem arises in the function `_check_start_shape` which passes the whole `start` dictionary (including the two original key-value pairs) leading to an exception on aesara side because those variables are not valid inputs. 

I was able to run this on my branch where I have refactored already the `DiscreteUniform`, which was the reason why this text was marked as an xfail, but it should be straightforward to get an example that is compatible with the current V4 branch.

brandonwillard commented:
> Sampling is failing when passing a dictionary for starting point with values for "untransformed" variables. The old test `TestDisasterModel` in `test_examples.py` that was marked as xfail shows this well:

So we can update the test to pass the transformed variables, or use/implement a function that converts a point `dict` with untransformed variables to transformed ones within `pm.sample`, no?

farhanreynaldo committed:
Remove GLM submodule in v4 (#4682)

* remove GLM submodule and its test

* add glm reference to Bambi

* add GLM removal to release notes

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>

ricardoV94 commented:
> So we can update the test to pass the transformed variables, or use/implement a function that converts a point `dict` with untransformed variables to transformed ones within `pm.sample`, no?

The `update_start_vals` function already converts transformed start vals to untransformed ones, so we just need a way to filter the original transformed values after this is done.

ricardoV94 committed:
Refactor several distributions (#4640)

* Refactor several distributions

* Fix continuous bounded default transform

* Add 32bit xfail to Weibull logp

* Add TODO reminder for Weibull

* Refactor random tests

* Remove tests covered by Aesara

* Refactor BetaBinomial

* Add skipif for betabinom depending on scipy version

* Rename transform_args to bound_args_indices

ricardoV94 committed:
Update dependency to aesara 2.0.8, and necessary fixes

ricardoV94 committed:
Use _logp and _logcdf dispatcher in ZeroInflated* methods

ricardoV94 committed:
Fix `check_logcdf` test regression

brandonwillard committed:
Revert "Automatically add SpecifyShape Op when full-length shape is given"

This reverts commit c99f15c28f5fd6635f38e0d130f49d410501bc9b.

brandonwillard committed:
Revert "Allow parametrization through either shape, dims or size"

This reverts commit ed29203f447c7b6d1687e31a3eda3da000c349df.

farhanreynaldo committed:
refactor pareto and laplace (#4691)

* refactor pareto
* refactor laplace
* Reintroduce `Pareto` default transform

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

ricardoV94 committed:
Fix BinaryMetropolis astep

ricardoV94 committed:
Fix _check_start_shape

ricardoV94 committed:
Update failing tests

themrzmaster committed:
Refactor Student T Distribution (#4694)

* feat: adapt student t
* Change default parameterization in terms of sigma and tweak tests

Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

michaelosthege committed:
Disallow "__sample__" as a dimension name

Co-authored-by: Thomas Wiecki <thomas.wiecki@gmail.com>

michaelosthege committed:
Add OS-independent #4652 regression tests

michaelosthege committed:
Make change_rv_size more robust

Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>
Co-authored-by: Thomas Wiecki <thomas.wiecki@gmail.com>

michaelosthege committed:
Introduce ShapeWarning

brandonwillard committed:
Remove duplicate pandas_to_array call in Model.register_rv

brandonwillard committed:
Make sure new size values are int64

Closes #4652.

brandonwillard committed:
Fix mistaken use of change_rv_size with new_size None

michaelosthege committed:
Edit misleading comment to be more accurate

michaelosthege committed:
Refactor test to use InferenceData

michaelosthege committed:
Implement backwards-compatble shape and Ellipsis-enabled dims

Co-authored-by: Thomas Wiecki <thomas.wiecki@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

farhanreynaldo committed:
refactor Moyal distribution (#4704)

* refactor Moyal distribution

* fixed ndim_params on moyal

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>

michaelosthege committed:
Add Ellipsis-support for the shape kwarg

farhanreynaldo committed:
refactor kumaraswamy (#4706)

* refactor kumaraswamy
* add logcdf method

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

farhanreynaldo committed:
Refactor Rice and Skew Normal distribution (#4705)

* refactor Skewnormal and Rice distribution
* add test for rice b and skewnorm tau params
* change default parameter to b
* Add float32 xfail to `test_rice`

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

themrzmaster committed:
Refactor LogitNormal (#4703)

MarcoGorelli committed:
let pandas_to_array take pandas Index

brandonwillard committed:
Update Aesara requirement to 2.0.9

brandonwillard committed:
Use aesara.tensor.atleast_1d in pymc3.aesaraf.change_rv_size

ricardoV94 committed:
Add testval to flaky test (#4707)

twiecki committed:
Update version string and add warning on import.

brandonwillard committed:
Incrementally update the RNG state in Model

ricardoV94 committed:
Refactor Flat and HalfFlat distributions (#4723)

* Refactor Flat and HalfFlat distributions

* Re-enable Gumbel logp test

* Remove redundant test

brandonwillard committed:
Convert RandomVariables to in-place during graph optimization

brandonwillard committed:
Introduce Model.initial_values and deprecate testval in favor of initval

brandonwillard committed:
Replace uses of testval with initval

twiecki committed:
Separate shape logic into a separate file (#4708)

Co-authored-by: Michael Osthege <m.osthege@fz-juelich.de>

brandonwillard committed:
Remove unnecessary gammaln and psi from pymc3.distributions.special

michaelosthege committed:
Use initial instead of testval internally

Co-authored-by: Oriol Abril-Pla <oriol.abril.pla@gmail.com>

brandonwillard committed:
Use the correct givens in Model.set_initval

michaelosthege committed:
Add workaround to make pylint pass until gp.predict is refactored
